src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidConstArrays.cs

```diff
@@ -0,0 +1,132 @@
+
+                        argumentOperation = invocationOperation.Arguments.FirstOrDefault();
+                        if (argumentOperation is not null)
+                        {
+                            if (argumentOperation.Children.First() is not IConversionOperation conversionOperation
+                                || conversionOperation.Operand is not IArrayCreationOperation arrayCreation)
+                            {
+                                return;
+                            }
+                            arrayCreationOperation = arrayCreation;
+                        }
+                        else // An invocation, extension or regular, has an argument, unless it's a VB extension method call
+                        {
+                            // For VB extension method invocations, find a matching child
+                            arrayCreationOperation = (IArrayCreationOperation)invocationOperation.Descendants()
+                                .FirstOrDefault(x => x is IArrayCreationOperation);
+                            if (arrayCreationOperation is null)
+                            {
+                                return;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        return;
+                    }
+
+                    if (argumentOperation is not null)
+                    {
+                        ITypeSymbol originalDefinition = argumentOperation.Parameter.Type.OriginalDefinition;
+
+                        // Can't be a ReadOnlySpan, as those are already optimized
+                        if (readonlySpanType is not null && originalDefinition.Equals(readonlySpanType))
+                        {
+                            return;
+                        }
+
+                        // Check if the parameter is a function so the name can be set to null
+                        // Otherwise, the parameter name doesn't reflect the array creation as well
+                        if (functionType is not null)
+                        {
+                            isDirectlyInsideLambda = originalDefinition.Equals(functionType);
+                        }
+                    }
+
+                    // Must be literal array
+                    if (arrayCreationOperation.Initializer.ElementValues.Any(x => x is not ILiteralOperation))
+                    {
+                        return;
+                    }
+
+                    Dictionary<string, string?> properties = new()
+                    {
+                        { "paramName", isDirectlyInsideLambda ? null : argumentOperation?.Parameter?.Name }
+                    };
+
+                    context.ReportDiagnostic(arrayCreationOperation.CreateDiagnostic(Rule, properties.ToImmutableDictionary()));
+                },
+                OperationKind.ArrayCreation,
+                OperationKind.Invocation);
```

Comment:
It's definitely minor but I am wondering if code wouldn't be clearer with 2 calls to `RegisterOperationAction` (one for each operation kind) with the 2 parts calling the same method for the common code.

Reply:
That would result in some duplicate code that isn't necessary to split up. I'm actually cleaning it up and adding more comments to clarify the control flow. Whether the operation is an array creation or an invocation, what is important is that we have an array creation involved with some sort of argument, and then we report a diagnostic from that array creation given it meets other criteria. Tell me what you think.